### ISSUE TITLE
`packComponents` method takes too much time for a bit big inputs

### ISSUE THREAD

**Issue by canbax** [iVis-at-Bilkent/cytoscape.js-layout-utilities#22]
[From this minimal reproducible sample](https://codepen.io/canbax/pen/VwKMOZR?editors=1111) you could see that it takes too much time to calculate. (180 seconds)

The graph in the example contains 4 connected components, 157 nodes, and 154 edges. I think it should handle it faster

**Comment by ugurdogrusoz** [comment]
@canbax Can you please confirm that this performance problem is on **incremental** layout?

**Comment by canbax** [comment]
Yes. When I click to incremental layout with tiling is false, the browser freeze for 3-5 seconds.

I'm attaching the performance profile file and my Cytoscape graph as JSON for the records. Note that files formats are txt since GitHub doesn't permit JSON file upload.

[cy-with-disconnecteds.txt](https://github.com/iVis-at-Bilkent/cytoscape.js-layout-utilities/files/6043118/cy-with-disconnecteds.txt)
[Profile-20210225T160647.txt](https://github.com/iVis-at-Bilkent/cytoscape.js-layout-utilities/files/6043119/Profile-20210225T160647.txt)

**Comment by canbax** [comment]
Also, note that when tiling is disabled it packs things like a circle. Doesn't pack with the desired aspect ratio

https://user-images.githubusercontent.com/8426741/109158590-a0acce00-7784-11eb-96e7-b2dad1ff4b72.mp4

**Comment by ugurdogrusoz** [comment]
@canbax Tiling is handled by individual layouts (has nothing to do with packing components). In the video above, tiling is disabled (that's not the expected layout with tiling).

**Comment by ugurdogrusoz** [comment]
Looks like these are potential performance improvement areas:
- `step` parameter defaults to a large number and is taken as the number of iterations to apply without checking for convergence
- check whether `auto` is calculated correctly and in a non-expensive manner
- we check intersection area with an expensive function without first checking whether two polygons corresponding to components intersect
- intersection area calculation function from the turf library is not specific to convex polygons (more general) and could be replaced with a more specialized and efficient one.

**Comment by ugurdogrusoz** [comment]
@onsah Can we please do some performance tests with say 5, 10, 20, and 100 component graphs? We should perform the tests for each such graph as follows (perform 3-5 times and take the average):
- Perform non-incremental layout (randomize: true) with packing enabled <= measure this
- Move about half of the nodes in the graph randomly between [0, 100] pixels
- Perform incremental layout (randomize: false) with packing enabled <= measure this
